services/repository/repository.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repository
+
+import (
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/notification"
+)
+
+// CreateRepository creates a repository for the user/organization.
+func CreateRepository(doer, owner *models.User, opts models.CreateRepoOptions) (*models.Repository, error) {
+	repo, err := models.CreateRepository(doer, owner, opts)
+	if err != nil {
+		if repo != nil {
+			if errDelete := models.DeleteRepository(doer, owner.ID, repo.ID); errDelete != nil {
+				log.Error("Rollback deleteRepository: %v", errDelete)
+			}
+		}
+
+		return nil, err
+	}
+
+	notification.NotifyCreateRepository(doer, owner, repo)
+
+	return repo, nil
+}
+
+// ForkRepository forks a repository
+func ForkRepository(doer, u *models.User, oldRepo *models.Repository, name, desc string) (*models.Repository, error) {
+	repo, err := models.ForkRepository(doer, u, oldRepo, name, desc)
+	if err != nil {
+		return nil, err
```

Comment:
If fork fails, don't we need to delete anything that might have been created? This check seems to be missing from the old code as well.

Reply:
done.